### PR TITLE
Feature[activity detail]: events list

### LINKED ISSUE #23: Add events list in the activity detail screen
That can be nice.
Maybe also a graph with only this activity?

### PR DESCRIPTION
**Pull request by ayshiff**
Closes #23

## Introduction

This is still a work in progress because I need to do some **adjustments** and **refactoring**.
I want to have your point of view on my implementation.

This Draft PR adds two new components to the activity detail screen: the `events list` and the `activity graph`.

I found the use of a *chronogram* interesting to get an overview of the different events from a chronological point of view.
We could also add the total for each day to get a more global overview of the day as on the "home" chart.

## Implementation

<p align="center">

</p>

<p align="center">

</p>
 
<p align="center">

</p>

**NOTE:** The selection of the activity type is above. 
I wanted to keep the graph and the list of events in the same place and knowing that the list of events can be long, I preferred to leave this part at the top.

## Improvements

For the chronogram, I wanted to know if some events could **overlap**. Currently we cannot see an event of the same type occurring while another one is also occurring.

A first idea I had was to **overlay** the different timelines.
The second idea was to be able to make the timelines **less opaque** to let the different events appear.

<p align="center">

</p>

One of the main improvements I would like to make is to be able to retrieve events as the user scrolls down (*endless scrolling*). Currently all the events are retrieved at the same time and this can cause **performance issues** if the user has a lot of events.

Apart from these two improvements, I will try to improve the **loading performance** of this screen, fix some **UI issues ** and do some **refactoring**.

I would like to have your opinion on my implementation to know if I am going in the right direction or not.

### CODE REVIEW

File: src/components/WeeklyBarChartDetail.res [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    ->Option.getWithDefault([])
+
+  let categoryId = settings->Calendars.categoryIdFromActivityTitle(activityTitle)
+
+  <View style>
+    <Spacer />
+    <SpacedView vertical=None>
+      <Text style={theme.styles["textLight2"]}>
+        {if todayFirst == startDate {
+          "Chronogram"
```
Comment (MoOx) [inline src/components/WeeklyBarChartDetail.res:37]: I would not show to the end user a technical term like this.

**Review by MoOx — COMMENTED** [review]
This looks very promising. I am globally to totally cool with the idea and implementation!

Here are some comments and idea:

- For the graph, I am curious to know the motivation about a horizontal graph. I was having in mind a thing similar to the one on the home, a bit like the one in Health app (see below)
IMG_0033E9FAFB57-1
- I don't think it's a huge problem, not sure if this adds a lot of value. What could be done maybe is to do a transparent/more opaque thing when someone tap on an event below: when a single even is "selected" we could highlight it on the graph above (but again, not sure about the real value of this). Not sure yet if visualising overlapping events if very interesting (but I am speaking for myself, maybe you feel very different about this).
- For the event start/end time, I would have put the it on a single line same line as the date like this: start - end. No color change, just a range with a dash. I found this format to be used in iOS Health app. This way, event duration bar can be more precise and use more width like on the home activity list
- For the event list, I would have only displayed current week (the home selected on the home), so this should be faster to show first render, and lazy load previous week if the person 1) scroll to this view 2) wait a bit like 2s, then I will start to lazy load previous week, so it fetching would not be an issue. 
- I understand that having all event could be nice, so again in iOS Health app (this is a reference for me, as well as iOS Screen Time views) there is often a "See more data" button that leads to a simpler unique flatlist. On this dedicated screen (or even a modal if there is no other screen after this one), we could imaging loading a single week, and try another week every second (with a spinner below to indicated more are coming - until screen is full of items - then you show this only if user scroll very fast and reach the end).

I am open to discuss about all of this, if you want we can use Discord or something else if you are more confortable (or we can continue here, that works for me).

**Comment by ayshiff** [comment]
- The main reason behind the horizontal graph is because I found it more intuitive from a UX point of view.
But actually, I think a vertical graph would be more consistent with the rest of the application. It seems to me that google calendar also uses a vertical timeline.

- From my point of view I find it weird to have the same type of event that can overlap with another event of the same type. For me this scenario shouldn't happen.

- I agree with that! I just saw that the iOS calendar uses this format, that's why.
<p align="center">

</p>

- I agree to display only the current week to the user and lazy-load the rest. If the user browses the home page and clicks on an activity that happened two weeks ago for example, wouldn't it be simpler to show him directly this week?

- I'm willing to start with the easiest solution (maybe the button to start with) and then iterate on it to improve the user experience. (lazy-load...)

- I agree to use discord to discuss some details but I'd like to keep the main discussion here since this contribution will be shared in [myopensourcejourney.com](https://myopensourcejourney.com/) and I'd like to keep a written record of the most important discussions if that's ok with you 👍

**Comment by MoOx** [comment]
I saw your website, that's a really great initiative, so let's stay here to discuss :) 

I am good for everything you said, thanks for some explanations!

> I agree to display only the current week to the user and lazy-load the rest. If the user browses the home page and clicks on an activity that happened two weeks ago for example, wouldn't it be simpler to show him directly this week?

Actually I was thinking that we show "current week" even.  Since you tap on an activity from a given week, we don't want the user to have a view saying "No events for this activity this week" (you get the idea). So this should reflect the week currently viewed on home screen.
Make it simple, then we will be able to iterate.

**Comment by ayshiff** [comment]
I updated my existing implementation to only display the events of the week the user is on. We have the same behaviour as the one of the home page with the "Top activities".
Instead of retrieving the list of events for the last 5 weeks, we retrieve (and display) only the events related to the week the user is on.

I think we can iterate on this first version 👍

**Review by MoOx — COMMENTED** [review]
It's a really great work. Thanks for doing this.

That said, this product is already slow (and I am ashamed of that, because I am responsible), and this PR does bring more performance issues (probably because code is similar to others screens - screens that already have perf issues).

I cannot merge this as it is until I fix the underlaying problems first (and then bring improvements in this PR).

For example if you open an activity detail, but swipe back during loading phase, you will have a 1/2s freeze where the animation stop in the middle (which is very frustrating).

And also, then events (I don't know why) looks really slow to appear (on my iPhone 12 Pro Max, and on this phone, I should not accept a single delay, as it will be horrible on poor android devices) and you see various state where the events are not yet totally visible. I would prefer a single render phase (or maybe 2 renders, since we may need available width from a layout event).

https://user-images.githubusercontent.com/157534/105828783-d3b44400-5fc3-11eb-879f-e842f0979308.mov

So depending on your time and interest in the project, I would prefer to improve this before merging (as master branch is my production/stable branch). If your time is limited I will totally merge this in a temporary branch and review the code when master runs faster (I am working on performance issues on android this days, I just bought more android phones to have more real world performances in front of my eyes).

I hope you understand my approach. Again, thanks for starting this features!

File: src/Calendars.res [hunk 2]
```diff
@@ -235,3 +235,21 @@ let makeMapCategoryDuration = (
         | false => 0
         }
   )
+
+let filterEventsByTitle = (
+  events: array<ReactNativeCalendarEvents.calendarEventReadable>,
+  title: string,
+) => events->Array.keep(evt => !(!(evt.title == title)))
```
[the remark this answers is not in the document]
Comment (ayshiff) [reply src/Calendars.res:242]: Yes you're right!
I must have gotten lost in my logic by looking at the `filterEvents` functions in `Calendars.res`.
I could probably even use one of these to filter my events (like `filterEventsByActivities`)

**Comment by ayshiff** [comment]
@MoOx Thanks for your review!
I totally agree with you.
I will iterate on this version to improve performance.
I first wanted to have a working version to show you and to know if the logic seemed good to you.

**Comment by MoOx** [comment]
I have update lots of code recently to improve performances. Some changes might cause you trouble with your changes, sorry for that. I will push some final changes and it should be better for perf. You can track #30 to be notified :)

**Comment by ayshiff** [comment]
Hi @MoOx !
Sorry to keep you waiting. I have integrated the changes you made in master.
I think I still have some performance issues, I'll try to fix it as soon as possible.

**Comment by ayshiff** [comment]
I created #36 as a follow-up of this PR.
I will convert this PR to draft to continue working on the Activity Chart.

**Pull request closed without merging**